feat: 문자열 계산기 구현 및 테스트 완료

nhs0912 opened the pull request:
문자열 계산기 구현 및 테스트 완료하였습니다! 리뷰 요청합니다!

nhs0912 committed:
feat: 문자열 계산기 구현 및 테스트 완료

hhlyn0328 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.regex.Pattern;
+
+public class StringAddCalculator {
+    private static final int ZERO = 0;
```
매직넘버의 상수화 하는 목적에는 두가지가 있을 것 같습니다 

**1. 매직 넘버의 의미를 알기 어렵다
2. 매직 넘버는 수정하기 어렵다**

0을 ZERO 라고 상수화하셨는데 0을 사용하는 모든 곳을 대체하는 것이 아닌 것으로 보여요
현재 사용된 곳을 보면, CustomDelimeter 의 위치를 나타내는 인덱스를 특정하고자 한 것 같은데
그 의미가 잘 나타나지 않는것 같고, 따라서 수정이 용이해보이지도 않는 것 같습니다 

조금 더 의미를 잘 나타낼 수 있도록 변경해보면 어떨까요? 
그리고 더 상수로 선언할 수 있는 문자열이나 매직넘버는 없는지 확인해볼까요? ㅎㅎ

nhs0912 replied:
그렇네요! 의미를 부여하는게 더 좋을거 같네요! 
그런데 한편으로는 상수의 개수가 계속 늘어난다면 그것도 보기가 힘들어질거 같다는 생각이 드는데 
혜린님은 상수 개수가 많을 때 어떻게 처리 하시는지요?? 
그릐고 상수의 개수가 어느정도여야 적당하다고 생각이 드나요??

hhlyn0328 replied:
개인적으로 상수의 숫자에는 연연하지 않은 것 같아요 

상수로 선언했을 때, 매직넘버나 문자열보다 가독성이 좋다고 생각되는 경우 (기본값, 패턴 등 정책을 담고 있는 경우),
중복 사용되는 경우에는 보통 사용하려고 노력하는 편인거 같아요 :)

hhlyn0328 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,62 @@
+
+    private static int addNumberByCustomDelimiter(String inputNumberText) {
+        // java.util.regex 패키지의 Matcher, Pattern import
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(inputNumberText);
```
`Effective Java` 의 `Item6 불필요한 객체 생성을 피하라` 에 아래와 같이 나와있습니다.
`Pattern은 입력받은 정규표현식으로 해당하는 유한 상태 머신(finite state machine)을 만들기 때문에 인스턴스 생성 비용이 높다.`

따라서 문자열마다 패턴을 생성하기보다는 패턴자체를 상수로 선언하면 어떨까요?

nhs0912 replied:
덕분에 이팩티브 자바 해당 챕터를 읽게 되었네요! 
계속해서 Pattern을 생성하고 있어서 불필요한 것 같아요~ 앞으로 이런 부분을 신경 써서 코딩 해야겠네요! 
감사합니다!

hhlyn0328 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringAddCalculator {
```
정적 메서드만 제공하는 클래스는 private 생성자를 사용하여 객체 생성을 막아볼까요?

nhs0912 replied:
private 생성자를 통해서 생성자를 막는 것이 프로그램 버그가 더 줄어서 좋을거 같네요!

hhlyn0328 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,62 @@
+    private static void isWrongNumber(String[] numbers) {
+        String regularExpression = "^[0-9]*$";
+        for (String number : numbers) {
+            if (!Pattern.matches(regularExpression, number)) {
```
`indent(들여쓰기) depth를 2단계에서 1단계로 줄여라.`

위의 요구사항을 지켜볼까요?

nhs0912 replied:
아! 요구사항을 안지키고 말았군요 ㅠㅠㅠ 
바로 반영 하였습니다! ㅋㅋㅋ

hhlyn0328 commented on src/test/java/calculator/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,55 @@
+
+public class StringAddCalculatorTest {
+    @Test
+    public void splitAndSum_null_또는_빈문자() {
```
테스트코드는 기능의 확인 목적도 있지만, 기능의 요구사항과 목적을 잘 설명해주기도 합니다.
왜 `assertThat(result).isEqualTo(0); ` 를 만족해야 테스트가 통과하는지 설명이 되면 좋을 것 같아요

메서드 이름이나 @DisplayName 에 설명할 수도 있고, 
int defaultValue = 0; 과 같이 0에게 의미를 부여해볼 수도 있을 것 같아요!

nhs0912 replied:
그렇네요! 기본 테스트 코드가 짜져 있어서 최소한 건드리진 않았는데 
저는 개인적으로 변수에다가 의미를 두는것이 제일 좋다고 생각이 들긴 하지만 테스트 코드에서는 @Display로 한글로 쓰여지는게 보는 사람을 위해서 좋을거 같아서 선호하고 있습니다! 만약 글로벌 환경이라면 달라지겠지만요 :) 
다른 테스트 코드에서도 설명 추가 하였어요!

hhlyn0328 reviewed (changes requested):
2단계 문자열 계산기 미션을 잘 진행해주셨네요!

소소한 코멘트 남겨보았습니다!
이번 미션 잘 부탁드려요 :)

nhs0912 committed:
refactor: 요구사항에 맞도록 수정

- 상수로 나타내는 부분 상수로 설정
- 상수에게 의미 부여하기
- 불필요한 객제 생성하지 않기
- static method로만 이루어진 클래스는 private 생성자 만들기

nhs0912 commented:
리뷰 코멘트 덕분에 공부가 잘 되었네요! ㅋㅋㅋ 앞으로 계속 조그만한 것이라도 부탁드려요! 
성장하고 싶어서요! ㅋㅋㅋ

hhlyn0328 reviewed (approved):
피드백 반영을 잘해주셨네요 👍 

다음 단계 진행을 위해 머지하겠습니다! 
다음 단계도 화이팅! 🔥🔥
